### ISSUE TITLE
AI

### ISSUE THREAD

**Issue by Showkun289** [Showkun289/index#3]
`Mini-GPT: โมเดลภาษาแบบเดียวกับ GPT/Claude แต่ย่อส่วนให้เล็กพอรันบนเครื่องส่วนตัวได้
สถาปัตยกรรม: Transformer Decoder (การ attend ตัวอักษรก่อนหน้าเพื่อทำนายตัวอักษรถัดไป)

วิธีใช้:
    1. pip install torch
    2. แก้ตัวแปร text ด้านล่างเป็นข้อความที่อยากให้โมเดลฝึกอ่าน (ยิ่งยาวยิ่งดี)
    3. รัน: python mini_gpt.py
    4. โมเดลจะฝึกแล้ว "แต่ง" ข้อความใหม่ให้ดูตอนท้าย

อ้างอิงแนวคิดจาก nanoGPT ของ Andrej Karpathy (อดีตนักวิจัย OpenAI)
"""

import torch
import torch.nn as nn
import torch.nn.functional as F

# ---------------------------------------------------------------------------
# 1) ข้อมูลฝึก — ใส่ข้อความอะไรก็ได้ตรงนี้ (ภาษาไทย/อังกฤษก็ได้)
# ---------------------------------------------------------------------------
text = """
สวัสดีครับ นี่คือตัวอย่างข้อความสำหรับฝึกโมเดลภาษาขนาดเล็ก
ยิ่งข้อความยาวและหลากหลาย โมเดลยิ่งเรียนรู้ได้ดีขึ้น
ลองแทนที่ข้อความนี้ด้วยหนังสือ บทความ หรือบทสนทนาที่คุณสนใจ
"""

chars = sorted(list(set(text)))
vocab_size = len(chars)
stoi = {ch: i for i, ch in enumerate(chars)}
itos = {i: ch for i, ch in enumerate(chars)}

def encode(s):
    return [stoi[c] for c in s]

def decode(ids):
    return "".join(itos[i] for i in ids)

data = torch.tensor(encode(text), dtype=torch.long)
n = int(0.9 * len(data))
train_data, val_data = data[:n], data[n:]

# ---------------------------------------------------------------------------
# 2) ไฮเปอร์พารามิเตอร์ (ปรับเพิ่มได้ถ้ามี GPU แรงๆ)
# ---------------------------------------------------------------------------
block_size = 32     # โมเดลมองย้อนหลังกี่ตัวอักษร (context length)
batch_size = 16      # จำนวนตัวอย่างต่อรอบฝึก
n_embd = 64          # ขนาดเวกเตอร์แทนแต่ละตัวอักษร
n_head = 4           # จำนวนหัว attention
n_layer = 4           # จำนวนชั้น transformer
dropout = 0.1
learning_rate = 3e-4
max_iters = 2000
device = "cuda" if torch.cuda.is_available() else "cpu"

def get_batch(split):
    d = train_data if split == "train" else val_data
    ix = torch.randint(len(d) - block_size, (batch_size,))
    x = torch.stack([d[i:i + block_size] for i in ix])
    y = torch.stack([d[i + 1:i + block_size + 1] for i in ix])
    return x.to(device), y.to(device)

# ---------------------------------------------------------------------------
# 3) สถาปัตยกรรม Transformer (หัวใจของ GPT/Claude)
# ---------------------------------------------------------------------------
class Head(nn.Module):
    """Self-attention หนึ่งหัว: แต่ละตัวอักษร 'มองย้อนกลับ' ไปตัวอักษรก่อนหน้าที่เกี่ยวข้อง"""
    def __init__(self, head_size):
        super().__init__()
        self.key = nn.Linear(n_embd, head_size, bias=False)
        self.query = nn.Linear(n_embd, head_size, bias=False)
        self.value = nn.Linear(n_embd, head_size, bias=False)
        self.register_buffer("tril", torch.tril(torch.ones(block_size, block_size)))
        self.dropout = nn.Dropout(dropout)

    def forward(self, x):
        B, T, C = x.shape
        k, q, v = self.key(x), self.query(x), self.value(x)
        wei = q @ k.transpose(-2, -1) * C ** -0.5
        wei = wei.masked_fill(self.tril[:T, :T] == 0, float("-inf"))  # ห้ามมองอนาคต
        wei = F.softmax(wei, dim=-1)
        wei = self.dropout(wei)
        return wei @ v

class MultiHeadAttention(nn.Module):
    def __init__(self, num_heads, head_size):
        super().__init__()
        self.heads = nn.ModuleList([Head(head_size) for _ in range(num_heads)])
        self.proj = nn.Linear(n_embd, n_embd)
        self.dropout = nn.Dropout(dropout)

    def forward(self, x):
        out = torch.cat([h(x) for h in self.heads], dim=-1)
        return self.dropout(self.proj(out))

class FeedForward(nn.Module):
    def __init__(self, n_embd):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(n_embd, 4 * n_embd),
            nn.ReLU(),
            nn.Linear(4 * n_embd, n_embd),
            nn.Dropout(dropout),
        )

    def forward(self, x):
        return self.net(x)

class Block(nn.Module):
    """หนึ่งชั้น Transformer: attention + feed-forward + residual connections"""
    def __init__(self, n_embd, n_head):
        super().__init__()
        head_size = n_embd // n_head
        self.sa = MultiHeadAttention(n_head, head_size)
        self.ffwd = FeedForward(n_embd)
        self.ln1 = nn.LayerNorm(n_embd)
        self.ln2 = nn.LayerNorm(n_embd)

    def forward(self, x):
        x = x + self.sa(self.ln1(x))
        x = x + self.ffwd(self.ln2(x))
        return x

class MiniGPT(nn.Module):
    def __init__(self):
        super().__init__()
        self.token_embedding = nn.Embedding(vocab_size, n_embd)
        self.position_embedding = nn.Embedding(block_size, n_embd)
        self.blocks = nn.Sequential(*[Block(n_embd, n_head) for _ in range(n_layer)])
        self.ln_f = nn.LayerNorm(n_embd)
        self.lm_head = nn.Linear(n_embd, vocab_size)

    def forward(self, idx, targets=None):
        B, T = idx.shape
        tok_emb = self.token_embedding(idx)
        pos_emb = self.position_embedding(torch.arange(T, device=device))
        x = tok_emb + pos_emb
        x = self.blocks(x)
        x = self.ln_f(x)
        logits = self.lm_head(x)

        if targets is None:
            loss = None
        else:
            B, T, C = logits.shape
            loss = F.cross_entropy(logits.view(B * T, C), targets.view(B * T))
        return logits, loss

    def generate(self, idx, max_new_tokens):
        """ทำนายตัวอักษรถัดไปทีละตัว แล้วป้อนกลับเข้าไปเรื่อยๆ (นี่คือวิธี GPT 'แต่งประโยค')"""
        for _ in range(max_new_tokens):
            idx_cond = idx[:, -block_size:]
            logits, _ = self(idx_cond)
            logits = logits[:, -1, 🙂
            probs = F.softmax(logits, dim=-1)
            idx_next = torch.multinomial(probs, num_samples=1)
            idx = torch.cat((idx, idx_next), dim=1)
        return idx

# ---------------------------------------------------------------------------
# 4) ฝึกโมเดล
# ---------------------------------------------------------------------------
if _name_ == "_main_":
    model = MiniGPT().to(device)
    print(f"จำนวนพารามิเตอร์: {sum(p.numel() for p in model.parameters())/1e3:.1f}K")

    optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)

    for step in range(max_iters):
        xb, yb = get_batch("train")
        logits, loss = model(xb, yb)
        optimizer.zero_grad(set_to_none=True)
        loss.backward()
        optimizer.step()

        if step % 200 == 0:
            print(f"step {step}: loss {loss.item():.4f}")

    # ---------------------------------------------------------------------
    # 5) ให้โมเดลแต่งข้อความใหม่ (นี่คือสิ่งเดียวกับที่ GPT/Claude ทำตอนตอบคุณ)
    # ---------------------------------------------------------------------
    print("\n--- ข้อความที่โมเดลแต่งเอง ---")
    context = torch.zeros((1, 1), dtype=torch.long, device=device)
    generated = model.generate(context, max_new_tokens=300)[0].tolist()
    print(decode(generated))`
